File: youzi_v2/routers/quote_opportunities.py
from fastapi import APIRouter, HTTPException, Query

from ..context import *
from ..schemas.quote_opportunities import (
    QuoteExtendDeadlineIn,
    QuoteFollowupIn,
    QuoteMarkLostIn,
    QuoteOpportunityCreateIn,
    QuoteOpportunityPatchIn,
)

router = APIRouter()


def _create_payload(body: QuoteOpportunityCreateIn) -> dict:
    return {
        "customer_id": body.customer_id,
        "customer_name": body.customer_name,
        "is_new_customer": body.is_new_customer,
        "customer_inquiry_no": body.customer_inquiry_no,
        "quote_date": body.quote_date,
        "deadline_date": body.deadline_date,
        "followup_interval_days": body.followup_interval_days,
        "next_followup_date": body.next_followup_date,
        "owner": body.owner,
        "product_name": body.product_name,
        "address_text": body.address_text,
        "ctns": body.ctns,
        "weight_kg": body.weight_kg,
        "volume_cbm": body.volume_cbm,
        "quoted_amount": body.quoted_amount,
        "quoted_currency": body.quoted_currency,
        "profit_amount": body.profit_amount,
        "profit_currency": body.profit_currency,
        "profit_rate": body.profit_rate,
        "note": body.note,
    }


@router.get("/api/v1/quote-opportunities/notifications/summary")
def quote_opportunity_notification_summary():
    return quote_opportunities_repo.notification_summary()


@router.get("/api/v1/quote-opportunities")
def list_quote_opportunities(
    scope: str = Query("todo"),
    status: str | None = None,
    customer: str | None = None,
    is_new_customer: bool | None = Query(None, alias="isNewCustomer"),
    owner: str | None = None,
    quote_date_from: str | None = Query(None, alias="quoteDateFrom"),
    quote_date_to: str | None = Query(None, alias="quoteDateTo"),
    deadline_from: str | None = Query(None, alias="deadlineFrom"),
    deadline_to: str | None = Query(None, alias="deadlineTo"),
    search: str | None = None,
    limit: int = 50,
    offset: int = 0,
):
    return quote_opportunities_repo.list_rows(
        scope=scope,
        status=status,
        customer=customer,
        is_new_customer=is_new_customer,
        owner=owner,
        quote_date_from=quote_date_from,
        quote_date_to=quote_date_to,
        deadline_from=deadline_from,
        deadline_to=deadline_to,
        search=search,
        limit=limit,
        offset=offset,
    )


@router.post("/api/v1/quote-opportunities")
def create_quote_opportunity(body: QuoteOpportunityCreateIn):
    try:
        return quote_opportunities_repo.create(_create_payload(body))
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/api/v1/quote-opportunities/{quote_id}")
def get_quote_opportunity(quote_id: str):
    row = quote_opportunities_repo.get_by_id(quote_id)
    if row is None:
        raise HTTPException(status_code=404, detail="报价不存在")
    return row


@router.patch("/api/v1/quote-opportunities/{quote_id}")
def patch_quote_opportunity(quote_id: str, body: QuoteOpportunityPatchIn):
    data = body.model_dump(exclude_unset=True)
    payload = {
        "customer_id": data.get("customer_id"),
        "customer_name": data.get("customer_name"),
        "is_new_customer": data.get("is_new_customer"),
        "customer_inquiry_no": data.get("customer_inquiry_no"),
        "quote_date": data.get("quote_date"),
        "deadline_date": data.get("deadline_date"),
        "followup_interval_days": data.get("followup_interval_days"),
        "next_followup_date": data.get("next_followup_date"),
        "status": data.get("status"),
        "owner": data.get("owner"),
        "product_name": data.get("product_name"),
        "address_text": data.get("address_text"),
        "ctns": data.get("ctns"),
        "weight_kg": data.get("weight_kg"),
        "volume_cbm": data.get("volume_cbm"),
        "note": data.get("note"),
        "lost_reason": data.get("lost_reason"),
    }
    payload = {k: v for k, v in payload.items() if k in data}
    try:
        return quote_opportunities_repo.update(quote_id, payload)
    except KeyError:
        raise HTTPException(status_code=404, detail="报价不存在") from None


@router.post("/api/v1/quote-opportunities/{quote_id}/followups")
def create_quote_followup(quote_id: str, body: QuoteFollowupIn):
    version = None
    if body.adjust_quote and body.version is not None:
        version = {
            "change_reason": body.version.change_reason,
            "product_name": body.version.product_name,
            "address_text": body.version.address_text,
            "ctns": body.version.ctns,
            "weight_kg": body.version.weight_kg,
            "volume_cbm": body.version.volume_cbm,
            "quoted_amount": body.version.quoted_amount,
            "quoted_currency": body.version.quoted_currency,
            "profit_amount": body.version.profit_amount,
            "profit_currency": body.version.profit_currency,
            "profit_rate": body.version.profit_rate,
            "note": body.version.note,
        }
        # 仅传入客户端显式字段，避免覆盖现有摘要
        version = {
            k: v
            for k, v in version.items()
            if v is not None or k in ("change_reason", "note")
        }
    try:
        return quote_opportunities_repo.create_followup(
            quote_id,
            followup_type=body.followup_type,
            note=body.note,
            next_followup_date=body.next_followup_date,
            adjust_quote=body.adjust_quote,
            version=version,
        )
    except KeyError:
        raise HTTPException(status_code=404, detail="报价不存在") from None
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/api/v1/quote-opportunities/{quote_id}/versions")
def list_quote_versions(quote_id: str):
    if quote_opportunities_repo.get_by_id(quote_id) is None:
        raise HTTPException(status_code=404, detail="报价不存在")
    return {"items": quote_opportunities_repo.list_versions(quote_id)}


@router.get("/api/v1/quote-opportunities/{quote_id}/followups")
def list_quote_followups(quote_id: str):
    if quote_opportunities_repo.get_by_id(quote_id) is None:
        raise HTTPException(status_code=404, detail="报价不存在")
    return {"items": quote_opportunities_repo.list_followups(quote_id)}


@router.post("/api/v1/quote-opportunities/{quote_id}/mark-won")
def mark_quote_won(quote_id: str):
    try:
        return quote_opportunities_repo.mark_won(quote_id)
    except KeyError:
        raise HTTPException(status_code=404, detail="报价不存在") from None


@router.post("/api/v1/quote-opportunities/{quote_id}/mark-lost")
def mark_quote_lost(quote_id: str, body: QuoteMarkLostIn):
    try:
        return quote_opportunities_repo.mark_lost(
            quote_id,
            lost_reason=body.lost_reason,
            note=body.note,
        )
    except KeyError:
        raise HTTPException(status_code=404, detail="报价不存在") from None


@router.post("/api/v1/quote-opportunities/{quote_id}/cancel")
def cancel_quote(quote_id: str):
    try:
        return quote_opportunities_repo.cancel(quote_id)
    except KeyError:
        raise HTTPException(status_code=404, detail="报价不存在") from None


@router.post("/api/v1/quote-opportunities/{quote_id}/extend-deadline")
def extend_quote_deadline(quote_id: str, body: QuoteExtendDeadlineIn):
    try:
        return quote_opportunities_repo.extend_deadline(
            quote_id,
            deadline_date=body.deadline_date,
            next_followup_date=body.next_followup_date,
        )
    except KeyError:
        raise HTTPException(status_code=404, detail="报价不存在") from None
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

File: youzi_v2/frontend/src/api/quoteOpportunities.ts
import { api } from '@/api/client'

export type QuoteScope =
  | 'todo'
  | 'today'
  | 'overdue_followup'
  | 'expiring_soon'
  | 'expired'
  | 'all_active'
  | 'won'
  | 'lost'
  | 'cancelled'

export interface QuoteOpportunity {
  id: string
  quoteNo: string
  customerId: string
  customerName: string
  isNewCustomer: boolean
  customerInquiryNo: string
  quoteDate: string
  deadlineDate: string
  followupIntervalDays: number
  nextFollowupDate: string
  status: string
  displayStatus: string
  statusLabel: string
  owner: string
  productName: string
  addressText: string
  ctns: number | null
  weightKg: number | null
  volumeCbm: number | null
  currentQuoteVersionId: string
  currentQuotedAmount: number | null
  currentQuotedCurrency: string
  currentProfitAmount: number | null
  currentProfitCurrency: string
  currentProfitRate: number | null
  lostReason: string
  note: string
  followupCount: number
  lastFollowupTime: string | null
  lastFollowupNote: string
  createdTime: string
  updatedTime: string
}

export interface QuoteVersion {
  id: string
  quoteId: string
  versionNo: number
  versionTime: string
  changeReason: string
  productName: string
  addressText: string
  ctns: number | null
  weightKg: number | null
  volumeCbm: number | null
  quotedAmount: number | null
  quotedCurrency: string
  profitAmount: number | null
  profitCurrency: string
  profitRate: number | null
  note: string
  createdTime: string
  updatedTime: string
}

export interface QuoteFollowup {
  id: string
  quoteId: string
  quoteVersionId: string
  followupTime: string
  followupType: string
  note: string
  nextFollowupDate: string
  createdBy: string
  createdTime: string
  updatedTime: string
}

export interface QuoteNotificationSummary {
  todayCount: number
  overdueCount: number
  expiringSoonCount: number
  pendingCount: number
}

export type QuoteCreateBody = {
  customerId?: string
  customerName: string
  isNewCustomer?: boolean
  customerInquiryNo?: string
  quoteDate?: string | null
  deadlineDate?: string | null
  followupIntervalDays?: number
  nextFollowupDate?: string | null
  owner?: string
  productName?: string
  addressText?: string
  ctns?: number | null
  weightKg?: number | null
  volumeCbm?: number | null
  quotedAmount?: number | null
  quotedCurrency?: string
  profitAmount?: number | null
  profitCurrency?: string
  profitRate?: number | null
  note?: string
}

export type QuoteFollowupBody = {
  followupType?: string
  note?: string
  nextFollowupDate?: string | null
  adjustQuote?: boolean
  version?: {
    changeReason?: string
    productName?: string
    addressText?: string
    ctns?: number | null
    weightKg?: number | null
    volumeCbm?: number | null
    quotedAmount?: number | null
    quotedCurrency?: string
    profitAmount?: number | null
    profitCurrency?: string
    profitRate?: number | null
    note?: string
  }
}

export async function listQuoteOpportunities(params?: {
  scope?: QuoteScope
  status?: string
  customer?: string
  isNewCustomer?: boolean
  owner?: string
  quoteDateFrom?: string
  quoteDateTo?: string
  deadlineFrom?: string
  deadlineTo?: string
  search?: string
  limit?: number
  offset?: number
}): Promise<{ items: QuoteOpportunity[]; total: number; limit: number; offset: number }> {
  const q: Record<string, string> = {}
  if (params?.scope) q.scope = params.scope
  if (params?.status?.trim()) q.status = params.status.trim()
  if (params?.customer?.trim()) q.customer = params.customer.trim()
  if (params?.isNewCustomer === true) q.isNewCustomer = 'true'
  if (params?.isNewCustomer === false) q.isNewCustomer = 'false'
  if (params?.owner?.trim()) q.owner = params.owner.trim()
  if (params?.quoteDateFrom?.trim()) q.quoteDateFrom = params.quoteDateFrom.trim()
  if (params?.quoteDateTo?.trim()) q.quoteDateTo = params.quoteDateTo.trim()
  if (params?.deadlineFrom?.trim()) q.deadlineFrom = params.deadlineFrom.trim()
  if (params?.deadlineTo?.trim()) q.deadlineTo = params.deadlineTo.trim()
  if (params?.search?.trim()) q.search = params.search.trim()
  if (params?.limit != null) q.limit = String(params.limit)
  if (params?.offset != null) q.offset = String(params.offset)
  return api('/api/v1/quote-opportunities', { query: q })
}

export async function getQuoteNotificationSummary(): Promise<QuoteNotificationSummary> {
  return api('/api/v1/quote-opportunities/notifications/summary')
}

export async function createQuoteOpportunity(body: QuoteCreateBody): Promise<QuoteOpportunity> {
  return api('/api/v1/quote-opportunities', { method: 'POST', body })
}

export async function getQuoteOpportunity(id: string): Promise<QuoteOpportunity> {
  return api(`/api/v1/quote-opportunities/${encodeURIComponent(id)}`)
}

export async function createQuoteFollowup(
  id: string,
  body: QuoteFollowupBody,
): Promise<QuoteFollowup> {
  return api(`/api/v1/quote-opportunities/${encodeURIComponent(id)}/followups`, {
    method: 'POST',
    body,
  })
}

export async function listQuoteVersions(id: string): Promise<{ items: QuoteVersion[] }> {
  return api(`/api/v1/quote-opportunities/${encodeURIComponent(id)}/versions`)
}

export async function listQuoteFollowups(id: string): Promise<{ items: QuoteFollowup[] }> {
  return api(`/api/v1/quote-opportunities/${encodeURIComponent(id)}/followups`)
}

export async function markQuoteWon(id: string): Promise<QuoteOpportunity> {
  return api(`/api/v1/quote-opportunities/${encodeURIComponent(id)}/mark-won`, {
    method: 'POST',
  })
}

export async function markQuoteLost(
  id: string,
  body: { lostReason?: string; note?: string },
): Promise<QuoteOpportunity> {
  return api(`/api/v1/quote-opportunities/${encodeURIComponent(id)}/mark-lost`, {
    method: 'POST',
    body,
  })
}

export async function cancelQuote(id: string): Promise<QuoteOpportunity> {
  return api(`/api/v1/quote-opportunities/${encodeURIComponent(id)}/cancel`, {
    method: 'POST',
  })
}

export async function extendQuoteDeadline(
  id: string,
  body: { deadlineDate: string; nextFollowupDate?: string | null },
): Promise<QuoteOpportunity> {
  return api(`/api/v1/quote-opportunities/${encodeURIComponent(id)}/extend-deadline`, {
    method: 'POST',
    body,
  })
}
